sdb/commands/linux/ps.py

```diff
@@ -0,0 +1,180 @@
+#
+# Copyright 2021 Datto Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pylint: disable=missing-docstring
+
+import datetime
+import argparse
+from textwrap import shorten
+from typing import Callable, Dict, Iterable, Union
+
+import drgn
+from drgn.helpers.linux.pid import for_each_task
+from drgn.helpers.linux.mm import cmdline
+
+import sdb
+from sdb.commands.internal.table import Table
+from sdb.commands.stacks import Stacks
+
+
+def _cmdline(obj: drgn.Object) -> str:
+    try:
+        s = " ".join(map(lambda s: s.decode("utf-8"), cmdline(obj)))
+
+        #
+        # The command line for a given thread can be obnoxiously long,
+        # so (by default) we limit it to 50 characters here. This helps
+        # preserve the readability of the command's output, but comes at
+        # the cost of not always showing the full command line of a
+        # thread.
+        #
+        return shorten(s, width=50)
+    except drgn.FaultError:
+        #
+        # The command line information is contained in the user address
+        # space of each thread, rather than in the kernel's address
+        # space. Thus, often, it may not be possible to retreive the
+        # thread's command line; e.g. when reading from a core dump.
+        #
+        return ""
+
+
+class Ps(sdb.Locator, sdb.PrettyPrinter):
```

Comment:
A lot of code in this class overlaps with `threads.py`. Instead of duplicating code in both places did you try the following:
[1] Extend the `threads` command to handle the extra columns that you added (e.g. `uid`) and also the column logic (e.g `-o` , `-h`, etc..)
[2] Make the `ps` command use the `threads` command under the hood and keep the arguments that make it look like the linux userland ps command (e.g. `-p`, `--ppid`, etc..).

If you want an example of layering a command on top of another command checkout the `spa` command. It is using  the `avl` walker for its Locator/no-input method, and `vdev` for its pretty-printer.

As for generic use of the `Table` class and how to implement the sort-by-column functionality you can look at either the `slubs` or `spl_kmem_caches`.

Reply:
Updated. Thanks for the advice.